src/hotspot/share/gc/serial/defNewGeneration.cpp

```diff
@@ -706,32 +700,27 @@ void DefNewGeneration::remove_forwarding_pointers() {
   // starts. (The mark word is overloaded: `is_marked()` == `is_forwarded()`.)
   struct ResetForwardedMarkWord : ObjectClosure {
     void do_object(oop obj) override {
-      if (obj->is_forwarded()) {
+      if (obj->is_self_forwarded()) {
```

Comment:
Why is `is_self_forwarded` treated specially? I'd expect the `is_forwarded` case alone to be enough here.

Reply:
Because I'd like self-forwarded marks not to be init-ed. Otherwise we'd have to preserve/restore them. Simply unset_self_forwarded() is enough to get them back to the original state.

Reply:
> Otherwise we'd have to preserve/restore them.

OK, then it's incorrect to use `init_mark()` for self-fwd objs.

This method is for self-fwd objs only. Can we remove the `else if` part? The non-self-fwd objs are essentially dead, right?

Reply:
I wasn't sure about this. What happens to successfully-promoted objects? Are we sure that no references point to their from-space parts? If yes, then I'd remove the else-if part and place an assert there instead.

Reply:
The else if part is needed when we later turn on UseCompactObjectHeaders, because the "normal" forwarding pointers then destroy the klass pointers, causing the object_iterate to fail to read the size of the objects.

Reply:
> What happens to successfully-promoted objects? ...

Successfully-forwarded objs are don't live in eden/from spaces, which are the spaces this closure is applied on. One can't have assert here, because as we iterate over eden/from spaces, we will encounter successfully-forwarded objs, but they should just be skipped.

> The else if part is needed when we later turn on UseCompactObjectHeaders...

I believe so, but it should be in the UseCompactObjectHeaders PR, not this one.

Reply:
Ok, thanks for the explanation!

I don't think it's needed for the UCOH part - the iterator fetched Klass*/size from forwardee if it encounters forwarded objects.

Reply:
You didn't change the iterators for the Serial GC, only for Parallel. But, yes, we can deal with this in the UCOH PR.